src/NLog/Internal/FilePathLayout.cs

```diff
@@ -164,6 +164,13 @@ private string GetRenderedFileName(LogEventInfo logEvent, System.Text.StringBuil
 
             if (reusableBuilder != null)
             {
+                if (!_layout.IsThreadAgnostic)
```

Comment:
It actually checks if there is a chance that the main-thread has already rendered the logevent using the layout. And if so then dives into the logevent-cache of pre-rendered-layout-results.